README
HTML::DOM, version 0.054

HTML::DOM is a Perl implementation of the HTML Document Object Model

This is an alpha release.  So far,  the level-2 core,  HTML and event
DOM interfaces have been implemented,  and some of the level-2  style
sheet interfaces.


RECENT CHANGES

See the Changes file for the long version.

0.054
-----
New feature:  The named node map returned by $element->attributes can
now be used as an array or hash reference.  This also means that Java-
Script code run by WWW::Scripter can do form.attributes[x].value, etc.

0.053
-----
Bug fixes:
• An extraneous  <body>  tag with  event  attributes  now  calls  the
  event_attr_handler on the body, not the preceding tag, and actually
  supplies the value of the attribute,  rather than undef (RT #75997).
  Thanks again to Tomki Camp for the report.
• DOMAttrModified  events now always have their type and  cancellabil-
  ity set correctly.

0.052
-----
Bug fixes:
• An extraneous  <body>  tag no longer  closes  currently  open  tags,
  turning <form><body><input> into  <form></form><input>  (RT #75997).
  Thanks to Tomki Camp for the report.
• The weird colour methods of the document (fgColor,  etc.) no longer
  die when there is no body element.


TO DO

- Finish Level 2 CSS support
- outerHTML/innerText?
- Other DOM interfaces (the rest of level 2, and level 3)
- HTML 5 stuff
- Write more complete documentation
- Write more tests
- Finishing checking for memory leaks (run all tests under
  Devel::Leak::Object)


INSTALLATION

The easiest way to install this module is to use the CPAN module or
the cpan script:

    [sudo] perl -MCPAN -e "install HTML::DOM"

    [sudo] cpan HTML::DOM

Or you can use the following:

    perl Makefile.PL
    make
    make test
    [sudo] make install


DEPENDENCIES

This module requires perl 5.8.3 or higher and the following
Perl modules:

- Scalar::Util 1.14 or later
- Exporter 5.57 or later
- URI
- LWP 1.13 or later
- CSS::DOM 0.07 or later
- HTML::Tagset 3.02 or later
- HTML::Parser 3.46 or later
- HTML::Encoding is required if the parse_file method is to be used.
- Tie::RefHash::Weak 0.08 or higher, if you are using perl 5.8.x


DOCUMENTATION

After installing, you can find documentation for this module with the 
perldoc command.

    perldoc HTML::DOM

Or try using man (it's faster, in my experience):

    man HTML::DOM


COPYRIGHT AND LICENCE

Copyright (C) 2007-14 Father Chrysostomos

This program is free software; you may redistribute it and/or modify
it under the same terms as perl.


Changes
Revision history for HTML::DOM

0.054   26 October, 2014
        New feature: The named node map returned by $elem->attributes
        can now be used as an array  or  hash  reference.  This  also
        means that  JavaScript  code  run  by  WWW::Scripter  can  do
        form.attributes['action'].value, etc.

0.053   April Fools’ Day, 2012
        Bug fixes:
        • An extraneous  <body>  tag with event attributes now  calls
          the event_attr_handler on the body, not the preceding tag’s
          element, and actually supplies the value of  the  attribute,
          rather than undef  (RT #75997).  Thanks again to Tomki Camp
          for the report.
        • When DOMAttrModified events were triggered, if it happened
          that no event listeners were triggered, the event object
          that HTML::DOM::EventTarget passed to the
          default_event_handler would not have its type set and its
          cancellability would be set to true instead of false. This
          has been fixed.

0.052   25 March, 2012
        Bug fixes:
        • An extraneous  <body>  tag no longer closes currently  open
          tags, turning <form><body><input> into <form></form><input>
          (RT #75997).  Thanks to Tomki Camp for the report.
        • The weird colour methods of the document (fgColor, etc.) no
          longer die when there is no body element.

0.051   11 March, 2012
        Bug fix:  write()  no longer dies inside an  element  handler
        (e.g.,  <script> element) if the document has no root element.
        Miscellaneous:
        • t/html-form.t  no  longer  fails  with  HTTP::Message  6.03
          (RT #75312).
        • This version drops the dependency on HTML::Tree.  HTML::DOM
          is trying to extend it in a way that  HTML::Tree  was never
          designed for. The result is a pile of hacks upon hacks that
          is very hard to manage. So forked versions of HTML::Element
          and HTML::TreeBuilder are now included  (whose forked names
          remain undocumented in case anyone is tempted to  use  them
          directly). This is the first step toward refactoring things
          and merging the  two  HTML  parsers  (what  we  essentially
          have) together.

0.050   1 January, 2012
        Bug fixes:
        • $doc->cookie  no longer  lists  cookies  from  the  request
          object in addition to those from the  cookie  jar  (usually
          resulting in duplicates).
        • HTML::DOM::Interface now makes sure that all modules provid-
          ing constants are loaded.

0.049   4 Sepember, 2011
        New feature:  <button>  elements are now treated as real  but-
        tons, instead of being ignored by most form-handling function-
        ality.  Thanks to a certain Martin for providing a patch upon
        which this was based (RT #70499).

0.048   27 March, 2011
        Bug fix:  Area elements now trigger a DOMActivate  event when
        the click method is called.  (This means that a default event
        handler for links gets called with area elements,  as well as
       ‘a’ elements.)

0.047   13 February, 2011
        Bug fix:  Implicitly-closed forms were being associated  with
        inputs in subsequent forms,  resulting  in  there  being  two
        forms that contain the same input element in their ->elements
        and ->inputs arrays (RT #65363).

0.046   23 January, 2011
        Bug fixes:
        • Setting innerHTML and calling the  insertAdjacent*  methods
          now cause node lists and collections to be updated.
        • Calling  ->checked(1)  on a radio button now unchecks other
          radio buttons in the same group.
        • Checking  a  radio  button  via  the  HTML::Form  interface
          ($form->value(...), $form->find_input(...)->value(...)) now
          causes other radio buttons in  the  same  group  to  become
          unchecked.

0.045   21 November, 2010
        Incompatible change:  event_parent now holds a weak reference.
        New feature: weaken_response constructor argument to indicate
        that the HTML::DOM object should hold a weak reference to the
        response object.

0.044   26 September, 2010
        Bug fixes:
        • Fixed to work with HTML::Tree 4

0.043   15 August, 2010
        Bug fixes:
        • Fixed to work with HTML::Tree 3.23_x
        • HTML::DOM::Element  can now  be  loaded  without  HTML::DOM
          already loaded.

0.042   18 July, 2010
        One bug fix:  Unmatched </td> and </th>  tags no longer close
        elements that  are  outside  the  innermost  enclosing  table.

0.041   13 June, 2010
        Bug fixes:
        • A CharacterData node’s data method now stringifies its argu-
          ment,  to avoid  tripping  up  HTML::Element  methods  like
          as_text by storing a reference in the attribute.  This also
          affects other methods,  including a  title  element’s  text
          method and a document’s title method.
        • Setting the selected attribute of an  orphaned  option  ele-
          ment  (one that is not inside  a  select  elem)  no  longer
          causes errors.
        • innerText is now listed in %HTML::DOM::Interface.
        • $element->getElementsByTagName  used to start searching  at
          the $element itself, instead of its children. Now it starts
          with the children.
        • $element->setAttribute  now  stringifies  the  value.  This
          solves the problem of getAttributeNode’s returning just the
          value instead of an Attr node,  if the value happened to be
          a reference.
        • cloneNode  with a true argument  (deep clone)  now sets the
          ownerDocument properly.
        • Setting a boolean attribute through the DOM 0 interface now
          works properly:  instead of being set to whatever value you
          pass in converted to a string, it is set to the name of the

Makefile.PL
BEGIN {
	require 5.008003; # substr fixes
}
use ExtUtils::MakeMaker;

# This little piece of code is to solve (at least for this module) the oft-
# repeated problem of upgrades breaking code that relies on it. Releasing
# working versions of the other code is not a complete solution; the user
# has to be told as well.
# Unfortunately, previous versions of the JS plugin wouldn’t compile with-
# out WWW::Scripter loaded already. So this is a mess.
%broken_deps = (
 WWW'Scripter                   => { breaks => 0.016, use => 0.017 },
 WWW'Scripter'Plugin'JavaScript => { breaks => 0.002, use => 0.003 }
);
while(my($mod,$vers) = each %broken_deps) {
 if(eval{require WWW'Scripter}
    and eval"require $mod" and !eval{VERSION $mod $$vers{use}}) {
  push @broken_deps, $mod;
 }
}

if(@broken_deps == 1) {
  my $mod = shift @broken_deps;
  my $vers = $broken_deps{$mod};
  $message = "WARNING: This distribution breaks compatibility with"
           ." $mod $$vers{breaks} and earlier. Please consider upgrading"
           ." to version $$vers{use}.";
  print "\n";
  write;
  print "\n";
}
elsif(@broken_deps) {
 print "
WARNING: This distribution breaks compatibility with the follow-
ing modules:
",
  map "  $_ $broken_deps{$_}{breaks} and earlier - use $broken_deps{$_}{use}\n",
     @broken_deps;
 print "\n";
}

format STDOUT =
^<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<~~
$message
.

my %lwp_prq = (
         HTTP::Message             => eval('require HTTP::Message;
                                            VERSION HTTP::Message') == 5.81
                                            ? 5.811 : 0,
         HTTP::Request             => 0,
         HTTP::Request::Common     => 0,
         HTTP'Headers              => 1.59, # comes with LWP 5.79; for the
        'HTTP::Headers::Util'      => 0,    # header_field_names method
# used only in the tests:
         HTTP'Response             => 0,
);
# also for the tests:
$lwp_prq{HTTP'Message} = 1.34 unless $lwp_prq{HTTP'Message}; # array ref
                                                             # headers
my %prq = (
         Carp                      => 0,
         constant                  => 1.03, # multiple at once
        'CSS::DOM'                 => 0.06, # compute_style
        'CSS::DOM::Style'          => 0.05, # modification_handler
                                            # and ::Style::parse’s
        'CSS::DOM::StyleSheetList' => 0,    # stringification
         Date'Format               => 0,
         Exporter                  => 5.57, # for use Exporter 'import'
	'HTML::Tagset'             => 0,
        'HTML::TreeBuilder'        => 0,
        'HTML::Element'            => 0,
	'HTML::Encoding'           => 0,
         HTML::Parser              => 3,
         overload                  => 0,
        'Scalar::Util'             => 1.09, # refaddr
         strict                    => 0,
         eval { require Hash::Util::FieldHash } ? () : (
          Tie'RefHash'Weak         =>  .08, # fieldhash
         ),
         URI                       => 0,
         warnings                  => 0,
# used only in the tests:
         CSS'DOM                   => 0.07, # !important ua rules
         Scalar::Util              => 1.14, # refaddr with ties
        'Test::More'               => 0,
	 utf8                      => 0,
);

my $need_lwp5837;
if($] < 5.0080071) {
 while(my($m,$v) = each %lwp_prq) {
  if(!eval{"require $m"} || !eval{$m->VERSION($v)}) {
   ++$need_lwp5837; last;
  }
 }
}
if($need_lwp5837) {
 $prq{'Bundle::LWP5_837'} = 0;
}
else {
 %prq = (%prq, %lwp_prq);
}

WriteMakefile(
    NAME                => 'HTML::DOM',
    AUTHOR              => 'Father Chrysostomos <sprout & cpan . org>',
    VERSION_FROM        => 'lib/HTML/DOM.pm',
    ABSTRACT_FROM       => 'lib/HTML/DOM.pm',
    PL_FILES            => {},
    PREREQ_PM     => \%prq,
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz',
                             TAR => 'COPYFILE_DISABLE=1 tar' },
    clean               => { FILES => 'HTML-DOM-*' },
);
                     
